lib/softirq.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2019, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * @file	softirq.h
+ * @brief	Soft Interrupt handling primitives for libmetal.
+ */
+
+#ifndef __METAL_SOFTIRQ__H__
+#define __METAL_SOFTIRQ__H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** \defgroup soft irq Interrupt Handling Interfaces
+ *  @{ */
+
+#include <metal/irq.h>
+
+/**
+ * @brief	metal_softirq_init
+ *
+ * Initialize libmetal soft IRQs controller
+ *
+ * @return 0 on success, or negative value for failure
+ */
+int metal_softirq_init();
+
+/**
+ * @brief	metal_softirq_dispatch
+ *
+ * Dispatch the pending soft IRQs
+ */
+void metal_softirq_dispatch();
+
+/**
+ * @brief	metal_softirq_allocate
+ *
+ * Allocate soft IRQs
+ *
+ * @param[in]  num number of soft irqs requested
+ * @return soft irq base for success, or negative value for failure
+ */
+int metal_softirq_allocate(int num);
+
```

Comment:
We need some API to allow sleeping until an interrupt occurs using WFI or similar instructions.

Reply:
why in the softirq allocation function?

Reply:
> We need some API to allow sleeping until an interrupt occurs using WFI or similar instructions.

Need abstraction here, don't call wfi directly. In RTOS context, one thread idle doesn't mean the whole system is idle.